### ISSUE TITLE
House-prices

### ISSUE THREAD

**Issue by Jeremy-Lii** [Jeremy-Lii/Jeremy#1]
import sys
from copy import deepcopy
from typing import Optional

import pandas as pd
from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.preprocessing import OneHotEncoder

num_columns = ["LotFrontage", "LotArea", "MasVnrArea", "BsmtFinSF1",
                "BsmtFinSF2", "BsmtUnfSF", "TotalBsmtSF", "1stFlrSF", 
                "2ndFlrSF", "LowQualFinSF", "GrLivArea", "BsmtFullBath",
                "BsmtHalfBath", "FullBath", "HalfBath", "BedroomAbvGr", 
                "KitchenAbvGr", "TotRmsAbvGrd", "Fireplaces", "GarageCars", 
                "GarageArea", "WoodDeckSF", "OpenPorchSF", "EnclosedPorch",
                "3SsnPorch", "ScreenPorch", "PoolArea", "MiscVal"]
cat_columns = ["MSSubClass", "MSZoning", "Street", "LotShape",
          "LandContour", "Utilities" ,"LotConfig", "LandSlope",
          "Neighborhood", "Condition1", "Condition2", "BldgType",
          "HouseStyle", "OverallQual", "OverallCond", "RoofStyle",
          "RoofMatl" , "Exterior1st", "Exterior2nd", "ExterQual", 
          "ExterCond", "Foundation", "BsmtQual", "BsmtCond",
          "BsmtExposure", "BsmtFinType1", "BsmtFinType2", "Heating",
          "HeatingQC", "CentralAir", "Electrical", "KitchenQual",
          "Functional", "FireplaceQu", "GarageType", "GarageFinish", 
          "GarageQual", "GarageCond" , "PavedDrive", "PoolQC", "Fence",
          "MiscFeature", "MoSold", "YrSold", "SaleType", "SaleCondition"]

class GrandTransformer(BaseEstimator, TransformerMixin):

    """
    A unified feature transformation:
    Each of the features will be transformed according to its "type"
    """
    
    def __init__(self, min_frequency: Optional[float] = None):

        """
        :param min_frequency: Specifies the minimum frequency below which a category will be considered infrequent.
        """

        self.min_frequency = min_frequency
        self.transformation_object = {}
        self.features_configuration = {'passthrough': num_columns,
                                       'onehot': cat_columns}

    def fit(self, X, y=None):
        
        for key, values in self.features_configuration.items():
            if key == 'onehot':
                ohe = OneHotEncoder(handle_unknown="ignore", min_frequency=self.min_frequency)
                ohe.fit(X[values])
                self.transformation_object[key] = ohe
                sys.stdout.write(f"\n**** {key}: {ohe.get_feature_names_out()} ****\n")
                # transformers.append((key, onehot_object(), values))
            elif key == 'passthrough':
                pass
            else:
                raise f"{key}: {values} not implement yet"

        return self

    def transform(self, X, y=None):
        X_transformed = deepcopy(X[self.features_configuration['passthrough']])

        for key, values in self.features_configuration.items():
            if key == 'onehot':
                data = self.transformation_object[key].transform(X[values])
                df = pd.DataFrame(data=data.toarray(),
                                  columns=self.transformation_object[key].get_feature_names_out(),
                                  index=X.index)
                X_transformed = X_transformed.join(df)

        return X_transformed

df = pd.read_csv("/Users/jeremy/Desktop/house-prices-advanced-regression-techniques/train.csv", index_col=0)

df.head(5)

pd.isnull(df).any()

import numpy as np
from sklearn.impute import SimpleImputer

# process nan
# https://scikit-learn.org/stable/modules/generated/sklearn.impute.SimpleImputer.html

# numerical
imp_num = SimpleImputer(missing_values=np.nan, strategy='mean')
# categorical
imp_cat = SimpleImputer(missing_values=np.nan, strategy='most_frequent')

from sklearn.model_selection import train_test_split

# train-test split for model regularization

df_train, df_eval = train_test_split(df, test_size=0.2, random_state=42)

y_train = df_train['SalePrice']
y_eval = df_eval['SalePrice']

# process train data
data_train_num = imp_num.fit_transform(df_train[num_columns])
df_train_num = pd.DataFrame(data=data_train_num, columns=num_columns)

data_train_cat = imp_cat.fit_transform(df_train[cat_columns])
df_train_cat = pd.DataFrame(data=data_train_cat, columns=cat_columns)

df_train = df_train_num.join(df_train_cat)

# process eval_data
data_eval_num = imp_num.transform(df_eval[num_columns])
df_eval_num = pd.DataFrame(data=data_eval_num, columns=num_columns)

data_eval_cat = imp_cat.transform(df_eval[cat_columns])
df_eval_cat = pd.DataFrame(data=data_eval_cat, columns=cat_columns)

df_eval = df_eval_num.join(df_eval_cat)

# df.filter(like='Lot')

## Transform all the features into numerical data

transformer = GrandTransformer()

x_train = transformer.fit_transform(df_train)
x_eval = transformer.transform(df_eval)

# Standardized the data

pd.isnull(x_train).any().sum()

for c in deepcopy(x_train.columns):
    if (x_train[c].std()) == 0:
        print(c)
        x_train.drop(c, inplace=True, axis=1)
        # x_eval.drop(c, inplace=True)

from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

x_train = scaler.fit_transform(x_train)
x_eval = scaler.transform(x_eval)

pd.isnull(x_train).any()

pd.isnull(y_train).any()

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

model = Sequential()
model.add(Dense(units=128, activation='relu', input_shape=(x_train.shape[1],)))
model.add(Dense(units=64, activation='relu'))
model.add(Dense(units=1, activation=None))  # 輸出層，不使用激活函數（線性輸出）

# 定義優化器，將學習率設置為較小的值（例如1e-4）
optimizer = Adam(learning_rate=1e-3)

# 編譯模型
model.compile(optimizer=optimizer, loss='mean_squared_error')

len(x_train)

import numpy as np

model.fit(x_train, np.log10(y_train), epochs=100, batch_size=32, validation_split=0.2)

predictions = model.predict(x_eval)

predictions

test_file_path = "/Users/jeremy/Desktop/house-prices-advanced-regression-techniques/test.csv"
df_test = pd.read_csv(test_file_path)
# process test_data
data_test_num = imp_num.transform(df_test[num_columns])
df_test_num = pd.DataFrame(data=data_test_num, columns=num_columns)

data_test_cat = imp_cat.transform(df_test[cat_columns])
df_test_cat = pd.DataFrame(data=data_test_cat, columns=cat_columns)

df_test = df_test_num.join(df_test_cat)

x_test = transformer.transform(df_test)

x_test = scaler.transform(x_test)

predictions = model.predict(x_test)
predictions = np.power(10, predictions)

submission_df = pd.DataFrame({'Id': df_test['Id'], 'SalePrice': predictions.flatten()})

# 將DataFrame保存為CSV文件
submission_df.to_csv('/Users/jeremy/Desktop/house-prices-advanced-regression-techniques/submission.csv', index=False)
